[aws-lambda] Add AsyncHandler

hansottowirtz opened the pull request:
This PR adds `AsyncHandler`, a subtype of `Handler`.

~This PR adds `PromiseBasedHandler`, `CallbackBasedHandler`, `ContextBasedHandler`, subtypes of `Handler`.~

Most modern AWS Lambda "handler wrapper" libraries export a `PromiseBasedHandler`.
e.g.:
- [@sentry/serverless](https://github.com/getsentry/sentry-javascript/blob/2794b6867a6971c2bad23f8b79544d4d7aff3a83/packages/serverless/src/awslambda.ts#L30)
- [@bugsnag/bugsnag-js](https://github.com/bugsnag/bugsnag-js/blob/0599faeb4019e571bfd7111c2663b5ddcbb78901/packages/plugin-aws-lambda/types/bugsnag-plugin-aws-lambda.d.ts#L6)
- [@vendia/serverless-express](https://github.com/vendia/serverless-express/blob/75cfbffeb3fd030695b9587a75300c7b8012ac6a/src/configure.js#L60)

They all define their own types for an `AsyncHandler` or similar. It would be great if we had a promise-only subtype for `Handler`, so we don't need to make our own types.

I added `AsyncHandler`. It is backwards compatible because I only added new type and ensured that it's compatible with `Handler`.

I know that there were already a lot of attempts to fix this, but I think this one is quite simple because it's completely backwards compatible and opt-in.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

hansottowirtz commented:
The `lambda-tester` tests also seem to fail on master.

simonbuchan reviewed (commented):
This is a lot more realistic an option than previous attempts, yes, but even with the test fixed I'm not sure how useful it will be as is. I expect most users of this library to be using the `FooHandler` type declared for each trigger, or just directly using the `FooEvent`, `FooResult`, and possibly the `Context`, directly on an async function. Do you have a use case for these types?

In any case, I would suggest removing the defaults for the generic parameters, or at least changing them to `unknown`. The point of this library is to have typed arguments and results, afterall!

With that, I would say this would be a lot nicer to use if you add for each trigger a define for the modern promise based handler type besides the legacy one (which, yes, probably should be `AsyncHandler`). I don't think the legacy formats are ones we want to be encouraging in that way, at least.

joeykilpatrick reviewed (commented):
Unfortunately these changes have the same problems as other attempts. A developer could technically implement a handler that satisfies these types, but then still have runtime errors if invoked by Lambda. A full description of this problem can be found here: https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/63474

hansottowirtz commented:
Thank you for the reviews. Generally, I think we need a migration path for developers and library authors to move from the legacy callback and context-based handlers to only promise-based handlers. Adding `AsyncHandler` here will make that easier, because there will be a standardized type.

I understand you might not think callback handlers are legacy, but I personally think they are already much less relevant than async handlers and are not considered idiomatic anymore. 

I have simplified the PR and now just export `AsyncHandler` next to `Handler`. This will be the most common use case by far anyway, so imo it doesn't make sense to export callback and context versions of `Handler`.

@simonbuchan

> In any case, I would suggest removing the defaults for the generic parameters, or at least changing them to unknown. The point of this library is to have typed arguments and results, afterall!

Agreed. I changed the generic parameters to `unknown`, I think this makes the most sense.

>  Do you have a use case for these types?

I think this will allow most authors of "wrapper" libraries to export an `AsyncHandler` instead of a `Handler`. The exported handlers are then easier to mock and test, because you know it only needs 2 arguments and will return a promise.

I personally just wrote a [small wrapper library](https://github.com/bubblydoo/wait-until-aws-lambda) for which I only want to support `AsyncHandler` based functions. So libraries like mine can then also restrict the handler that they wrap to be an `AsyncHandler`.

I then hope that in the future all libraries and projects will be using `AsyncHandler`.

> I would say this would be a lot nicer to use if you add for each trigger a define for the modern promise based handler type besides the legacy one

You mean adding `APIGatewayProxyAsyncHandler`, `AppSyncAuthorizerAsyncHandler` etc, correct?

@joeykilpatrick 
Thank you for that discussion, I hadn't seen it yet.
I have added a test for the `handler4` case, as you can see passing a callback to an `AsyncHandler` is not possible.
I think `AsyncHandler` does not need to be a problem. If you pass around `AsyncHandler`s, you know that they need 2 arguments and return a promise. If you pass around `Handler`s, you know they need 3 arguments and might return void or a promise, and you need to add logic for all 3 types of handlers.
I may be misunderstanding you here, please let me know if that is the case.

(I will also add better docs for the `AsyncHandler` later.)

hansottowirtz commented:
@joeykilpatrick I see the `newHandler7` indeed still fails. I'll investigate further.

joeykilpatrick commented:
Admittedly, unlike some previous PRs that change `Handler` to be a union, this PR _does_ prevent developers from using the callback when writing an async handler. 

However, here's a reasonable use case that still behaves unexpectedly:

```typescript
const myHandler = async (event: unknown, context: unknown, isTest?: boolean) => {

  if (isTest) {
    // Do test things
  } else {
    // Do prod things
  }

}


// Running my tests here
const event: unknown = {};
const context: unknown = {};
myHandler(event, context, true);
// ...
// ...


// Despite satisfying the type, Lambda will always pass a callback.
// This will unexpectedly pass a callback function for the 'isTest' 
// parameter causing the handler to behave unexpectedly. 
export const handler: AsyncHandler = myHandler;

```
[TS Playground Link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAwg9gO2BAHsKBeKBXBBrBOAdwQG4oB6CqAcQgQgCcBDAGwEsAvdhAcymAALdgGcB4aADM4jKCPYBbMBwDG7UADoAsAChdoSFACCIkAhUAJZggAmrJgB4AKgFEAbvXRZcBYggA0UE4AShAi2KxeOPiEJAB8mFAAFLpQaVAQHkgAXEHunv6p6SqIyGi58EiowIU6AJSYCQAKjHAKohDOoeGRcaS6AzolCCLoCiBWtvayWMym5smZnrk+sQFQw2XAKzF+gaJOYcAA-LkARnBw9tYNGAkA3oNp7JLJB0cNjzrplNQAInABEcBMI+CIigBfDKsETQL4-KhQAFQMCtGwgni8cHfKAQwZ4vSExHBXAITFQcZA0ZiQRMCC6YajDJZbbRXwkRL3CH9IaIJmbao7dkITnc3TjSZ2JhJJZIQICtCBYCMbAQOo8xEaLW6TXawk6-5hMDqaAiZjAUSSEDkoTQAwQQIAGWYCjONmYUCI7FYrCgbCIzBAYjAczEHpUbFYZ2YKjw2h0iKcwjEXp90VQkBUyDsIBRob9G0j0djUEkuCz7EQpZkIOgAHJ3qM61ADXmWAoIMhZBHsPI+LWoIJrFLZMBAWcIEOPOmUJns6wQBoWzoMzJ0Iz0EOpkxciYzJZh9NEhLD0weUA)

This PR and the ones before it aim to _narrow_ the definition of a handler to prevent the use of a callback when using an async handler. That is a great goal. Using the provided callback in an async handler is certainly a terrible practice and should be restricted if possible.

However, all of these attempts actually _widen_ the handler definition to include handlers that are not correct. At best they cause unintended behavior, at worst they cause runtime errors. The bar for including these type of changes in any DefinitelyTyped package is extraordinarily high. In my opinion, enforcing a best practice doesn't meet that standard.

Personally, I am always appreciative for those who are willing to take the time to submit a PR to make changes to these libraries to help others. This particular issue is tricky and unintuitive. I would be in favor of adding comments next to the `Handler` definition that explain the history of these past attempts.

hansottowirtz commented:
I think I have been able to prevent this use case from happening, by using an intersection:

```ts
export type AsyncHandler<TEvent = unknown, TResult = unknown> =
  ((event: TEvent, context: Context, callback: Callback<TResult>) => Promise<TResult>) &
  ((event: TEvent, context: Context) => Promise<TResult>);
```

This is kind of a self-contradicting type:
If you directly assign a function to an `AsyncHandler` you cannot have a 3rd parameter.
If you indirectly assign and you want to have a 3rd parameter it needs to be a `Callback`.

So I think this is a possible solution to this problem. Unfortunately it only works on Typescript >= 4.3, otherwise the types of `event` and `context` aren't correctly inferred.

Would you agree that, if I found a way to make these tests pass on Typescript >= 3.0, it would be worthwhile to continue the PR?

joeykilpatrick commented:
This appears to be a clever new approach to this long-running issue. 

If we can get this to work, then lets explore what the consequences would be for changing the `Handler` definition to 

```typescript
type Handler<TEvent = unknown, TResult = unknown> =  
    | ((event: TEvent, context: Context, callback: Callback<TResult>) => void)
    | AsyncHandler<TEvent, TResult>
    ;
```
